docs/tutorials/typescript/workqueue-slackbot/code/bot/modules/workqueue.ts

```diff
@@ -0,0 +1,390 @@
+import "dotenv/config";
+import {
+  SayFn,
+  RespondFn,
+  SlackCommandMiddlewareArgs,
+  GenericMessageEvent,
+} from "@slack/bolt";
+import {WebClient} from "@slack/web-api";
+import {WorkqueueData, WorkqueueStatus} from "../../common-types/types";
+import {temporalClient} from "./temporal";
+import crypto from "crypto";
+import {formatDistanceToNow} from "date-fns";
+
+// Handles and routes all incoming Work Queue messages
+export async function handleWorkqueueCommand(
+  command: SlackCommandMiddlewareArgs["command"],
+  client: WebClient,
+  say: SayFn,
+  respond: RespondFn
+) {
+  const commandText = command.text?.trim();
+
+  if (commandText === "!delete") {
+    await deleteWorkqueue(command, say, client);
+  } else if (commandText === "") {
+    await displayWorkqueue(command, respond, client);
+  } else {
+    await addWorkToQueue(command, say, client);
+  }
+  return;
+}
+
+// Handles the action of claiming a work item
+
+// Custom reply function that sends a message in the same thread
+// Defaults to text, but capable of sending blocks as well
+async function reply(say: SayFn, text: string, blocks?: any) {
+  try {
+    // If the message is already part of a thread, reply in the same thread
+    await say({
+      text, // Always include the text field
+      blocks, // Include blocks if provided
+    });
+  } catch (error) {
+    console.error("Error occurred while sending message:", error);
+    throw new Error("Failed to send message");
+  }
+}
+
+async function replyEphemeral(respond: RespondFn, text: string, blocks?: any) {
+  try {
+    await respond({
+      text, // Always include the text field
+      blocks, // Include blocks if provided
+    });
+  } catch (error) {
+    console.error("Error occurred while sending ephemeral message:", error);
+    throw new Error("Failed to send ephemeral message");
+  }
+}
+
+// Display the Work Queue for the channel
+// Creates a new Work Queue if it does not exist
+async function displayWorkqueue(
+  command: SlackCommandMiddlewareArgs["command"],
+  respond: RespondFn,
+  client: WebClient
+) {
+  // Get the channel name in plain text
+  const channelName = command.channel_name;
+  // Check if the Work Queue already exists for the channel
+  if (await checkIfWorkqueueExists(channelName)) {
```

Comment:
There's some possibility of race condition here. `checkIfWorkqueueExists` may return false, but then a workflow get started at the same time, so `createNewWorkqueue` fails. The opposite is also true (ie. `checkIfWorkqueueExists` returns true, but then, the workflow completes before you get to `queryWorkqueue`); that will not cause a failure (query a terminated workflow is legit), but based on the fact that `checkIfWorkqueueExists` assert that the workflow is running, it looks like querying a completed workflow is not the behavior you expect.

Instead, just try to create the queue directly, without calling `DescribeWorkflow` first, and simply ignore the `WorkflowExecutionAlreadyStartedError` exception that will be thrown if the Workflow already exists. For the query… Why is it important that the workflow still exists?

Reply:
Originally, I was thinking I didn't want to return data from an older Cancelled Work Queue.
But I suppose that is not bad.